source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc

```diff
@@ -178,25 +191,30 @@ UdpProxyFilter::StickySessionClusterInfo::StickySessionClusterInfo(
                                      HeterogeneousActiveSessionEqual(false))) {}
 
 Network::FilterStatus UdpProxyFilter::StickySessionClusterInfo::onData(Network::UdpRecvData& data) {
+  bool defer_socket = filter_.config_->hasSessionFilters();
   const auto active_session_it = sessions_.find(data.addresses_);
   ActiveSession* active_session;
   if (active_session_it == sessions_.end()) {
-    active_session = createSession(std::move(data.addresses_));
+    active_session = createSession(std::move(data.addresses_), nullptr, defer_socket);
     if (active_session == nullptr) {
       return Network::FilterStatus::StopIteration;
     }
   } else {
     active_session = active_session_it->get();
-    if (active_session->host().coarseHealth() == Upstream::Host::Health::Unhealthy) {
+    // We defer the socket creation when the session includes filters, so the filters can be
+    // iterated before choosing the host, to allow dynamically choosing upstream host. Due to this,
+    // we can't perform health checks during a session.
+    if (!defer_socket &&
+        active_session->host().value().get().coarseHealth() == Upstream::Host::Health::Unhealthy) {
```

Comment:
Wouldn't you want this behavior to happen at the end of the session filter chain though? Or give filters the chance to interact with this behavior? This seems like a regression in functionality for those cases. At minimum I would add some kind of TODO about this?

/wait

Reply:
I see your point. It makes sense that a similar check can also happen when using filters as you suggested. I added a TODO and will add this in a separate PR, need to do some more thinking about this. Thanks